### PR TITLE
chore: Write a mongo migration to give all existing users a default display name

### PR DESCRIPTION
**Pull request by jcbcapps**
# SC-1216

## Proposed changes
Adds a migration to give all existing users a default `displayName` field, if they do not already have one.

---

## Reviewer notes
You can run this migration against your local db. I'd suggest making sure that you have a few users and delete the `displayName` field for each. Here are the [migration docs](https://github.com/USSF-ORBIT/ussf-portal-client/blob/main/docs/development.md#database-migrations) if you'd like more info, but long story short:
- Run `yarn migrate up 1665502977143-add-display-name-to-user`. If this isn't working for some reason, you may need to run `yarn migrate up` and that will run ALL migrations. There is just one other migration, but still doesn't seem ideal to have to run it.
- Check your local db at `http://localhost:8888/` and verify that all users now have a `displayName` field with an empty string as the value.
- Optional: run `yarn migrate down 1665502977143-add-display-name-to-user` to rollback the migration. Same as above, if this isn't working just run `yarn migrate down` to run all migrations.
- Make sure unit tests are passing

## Setup
The usual:
- Run `yarn services:up && yarn dev` in the CMS repo
- Run `yarn dev` in the portal client repo

---

### As the original developer, I have

- [ ] Met the acceptance criteria, or will meet them in subsequent PRs or stories
  - ... 
- [ ] Created new stories in Storybook if applicable
  - [ ] Checked that all Storybook accessibility checks are passing
- [ ] Created/modified automated unit tests in Jest
  - [ ] Including jest-axe checks when UI changes
- [ ] Created/modified automated E2E tests in Cypress
  - [ ] Including cypress-axe checks when UI changes
  - [ ] Checked that the E2E test build is not failing
- Performed [a11y testing](https://github.com/trussworks/accessibility/blob/master/sample_a11y_testing_process.md):
  - [ ] Checked responsiveness in mobile, tablet, and desktop
  - [ ] Checked keyboard navigability
  - [ ] Tested with [VoiceOver](https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts) in Safari
  - [ ] Checked VO's [rotor menu](https://github.com/trussworks/accessibility/blob/master/README.md#how-to-use-the-rotor-menu) for landmarks, page heading structure and links
  - [ ] Used a browser a11y tool to check for issues (WAVE, axe, ANDI or Accessibility addon tab for Storybook)
- [ ] Requested a design review for user-facing changes
- For any new migrations/schema changes:
  - [ ] Followed guidelines for zero-downtime deploys

### As code reviewer(s), I have

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
  - [ ] Checked that the E2E test build is not failing
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] Considered marking this as accepted even if there are small changes needed
- Performed [a11y testing](https://github.com/trussworks/accessibility/blob/master/sample_a11y_testing_process.md):
  - [ ] Checked responsiveness in mobile, tablet, and desktop
  - [ ] Checked keyboard navigability
  - [ ] Tested with [VoiceOver](https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts) in Safari
  - [ ] Checked VO's [rotor menu](https://github.com/trussworks/accessibility/blob/master/README.md#how-to-use-the-rotor-menu) for landmarks, page heading structure and links
  - [ ] Used a browser a11y tool to check for issues (WAVE, axe, ANDI or Accessibility addon tab for Storybook)

### As a designer reviewer, I have

- [ ] Checked in the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow
- Performed [a11y testing](https://github.com/trussworks/accessibility/blob/master/sample_a11y_testing_process.md):
  - [ ] Checked responsiveness in mobile, tablet, and desktop
  - [ ] Checked keyboard navigability
  - [ ] Tested with [VoiceOver](https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts) in Safari
  - [ ] Checked VO's [rotor menu](https://github.com/trussworks/accessibility/blob/master/README.md#how-to-use-the-rotor-menu) for landmarks, page heading structure and links
  - [ ] Used a browser a11y tool to check for issues (WAVE, axe, ANDI or Accessibility addon tab for Storybook)

### As a test user, I have

- Run through the [Test Script](hhttps://docs.google.com/spreadsheets/d/1eV8UK0aJZ0qrzjnXf5SJ_uRiV7bpsj3yrhRFEkjOvV4/edit?usp=sharing):
  - [ ] On commercial internet in IE11
  - [ ] On commercial internet in Firefox
  - [ ] On commercial internet in Chrome
  - [ ] On commercial internet in Safari
  - [ ] On NIPR in IE11
  - [ ] On NIPR in Firefox
  - [ ] On NIPR in Chrome
  - [ ] On NIPR in Safari
  - [ ] On a mobile device in Firefox
  - [ ] On a mobile device in Chrome
  - [ ] On a mobile device in Safari
- [ ] Added any anomalous behavior to this PR

---

### CODE REVIEW

File: migrations/1665502977143-add-display-name-to-user.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const { runMigration } = require('../utils/mongodb')
+
+module.exports.up = runMigration(async (db) => {
+  await db.collection('users').updateMany(
+    {},
+    {
+      $set: {
+        displayName: '',
```
Comment (jcbcapps) [inline migrations/1665502977143-add-display-name-to-user.js:11]: I left this as just an empty string because the `<PersonalData />` component has the user context, and performs a check to see if there is a `displayName`. If there isn't, it provides the `givenname` and `surname` fields to display instead.

File: migrations/1665502977143-add-display-name-to-user.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      },
+    }
+  )
+})
+
+module.exports.down = runMigration(async (db) => {
+  await db.collection('users').updateMany({}, { $unset: { displayName: '' } })
```
Comment (gidjin) [inline migrations/1665502977143-add-display-name-to-user.js:18]: issue: this down migration will remove displayName even if it is not empty string

Suggestion: update the `updateMany` call to only apply if `displayName` is set to empty string.
Suggestion: add a couple of tests  that checks display name is not over written if set and not removed if not empty string

I point this out since I don't want to delete someone's customized name if they have selected it already if we ever need to run the down version of this migration. It could be that this is unnecessary but feels like good habit to be in.

**Review by gidjin — CHANGES_REQUESTED** [review]
I found one issue, I would be happy to pair on fixing it if that would be helpful.

File: migrations/1665502977143-add-display-name-to-user.js (hunk 2 above)
Comment (jcbcapps) [reply migrations/1665502977143-add-display-name-to-user.js:18]: Great catch! I think I could just apply a filter like in the `up` function to remedy this. I'll add those tests in as well.

**Review by gidjin — APPROVED** [review]
LGTM! once the e2e test data is fixed this is good to go!
